File: java/spoj/StockMax.java
package spoj;

import java.util.Scanner;

public class StockMax {
	static int p[] = new int[50000];
	public static int maxProfit(int[] arr,int index,int stocks,int profits) {
		if(index>=arr.length-1) {
			int p_sell=profits+stocks*arr[index];
			int p_nothing=profits;
			p[index]= Math.max(p_sell, p_nothing);
			return p[index];
		}
		
		if(p[index]!=Integer.MIN_VALUE) {
			return p[index];
		}
		
		int p_buy = maxProfit(arr, index+1,stocks+1,profits-arr[index]);
		
		int p_sell=maxProfit(arr, index+1, 0, profits+stocks*arr[index]);
		
		int p_nothing=maxProfit(arr, index+1, stocks, profits);
		
		p[index]= Math.max(p_buy, Math.max(p_sell, p_nothing));
		return p[index];
	}
	
	
	
	public static void main(String[] args) {
		Scanner s =new Scanner(System.in);
		int t=s.nextInt();
		for(int test=0;test<t;test++) {
			int n=s.nextInt();
			int arr[]=new int[n];
			for(int i=0;i<n;i++) {
				arr[i]=s.nextInt();
				p[i]=Integer.MIN_VALUE;
			}
			System.out.println(maxProfit(arr, 0, 0,0));
		}
	}
}

File: python/StockMax.py
t=int(raw_input())

dict={}

def maxProfit(arr,index,stocks,profits):
    if index==(len(arr)-1):
        p_sell=profits+stocks*arr[index];
        p_nothing=profits;
        dict[(index,stocks,profits)] = max(p_sell, p_nothing);
        return dict.get((index,stocks,profits))
    
    if dict.has_key((index,stocks,profits)):
        return dict.get((index,stocks,profits))    
    s=stocks+1
    p=profits-arr[index]    
    p_buy = maxProfit(arr, index+1,s,p)
    
    s=0
    p=profits+stocks*arr[index]
    p_sell=maxProfit(arr, index+1,s,p )
    
    p_nothing=maxProfit(arr, index+1, stocks, profits)
    
    dict[(index,stocks,profits)]=max(p_buy,p_sell, p_nothing)
    return dict.get((index,stocks,profits))
        
for test in range(t):
    n=int(raw_input())
    dict={}
    arr=[int(i) for i in raw_input().split(' ')]
    #print arr
    print maxProfit(arr, 0, 0, 0)
    

